libvast/src/system/node.cpp

```diff
@@ -575,68 +579,76 @@ node(node_actor::stateful_pointer<node_state> self, std::string name,
     VAST_DEBUG("{} got EXIT from {}", self, msg.source);
-    std::vector<caf::actor> scheduled_for_teardown;
-    // TODO: A recent refactoring introduced a bug that caused this function to
-    // only remove by label instead of by type or label. This was not caught in
-    // our integration test suite because we do not currently test continuous
-    // imports/exports in them. We should evaluate how we can test for shutdown
-    // bugs.
-    auto schedule_teardown = [&](const std::string& type_or_label) {
-      if (is_singleton(type_or_label)) {
-        if (auto component = self->state.registry.remove(type_or_label)) {
-          VAST_VERBOSE("{} schedules {} for shutdown", self, type_or_label);
-          self->demonitor(component->actor);
-          scheduled_for_teardown.push_back(std::move(component->actor));
-        }
-      } else if (auto components
-                 = self->state.registry.find_by_type(type_or_label);
-                 !components.empty()) {
-        VAST_VERBOSE("{} schedules {} {}(s) for shutdown", self,
-                     components.size(), type_or_label);
-        for (auto& component : components) {
-          if (auto removed = self->state.registry.remove(component)) {
-            self->demonitor(removed->actor);
-            scheduled_for_teardown.push_back(std::move(removed->actor));
-          }
-        }
+    self->state.tearing_down = true;
+    // Ignore duplicate EXIT messages except for hard kills.
+    self->set_exit_handler([=](const caf::exit_msg& msg) {
+      if (msg.reason == caf::exit_reason::kill) {
+        VAST_WARN("{} received hard kill and terminates immediately", self);
+        self->quit(msg.reason);
+      } else {
+        VAST_DEBUG("{} ignores duplicate EXIT message from {}", self,
+                   msg.source);
       }
-    };
-    // Terminate the accountant first because it acts like a source and may
-    // hold buffered data.
-    schedule_teardown("accountant");
-    // Take out the filesystem, which we terminate at the very end.
-    auto filesystem = deregister_component(self, "filesystem");
-    // Tear down the pipeline from source to sink. Note that the order is
-    // important here; the source must be shut down before the importer.
-    auto pipeline = {"source",  "importer", "disk-monitor", "index",
-                     "archive", "exporter", "sink"};
-    for (const auto* component : pipeline)
-      schedule_teardown(component);
-    // Now schedule all remaining components for termination.
-    auto& registry = self->state.registry;
-    std::vector<std::string> remaining;
-    remaining.reserve(registry.components().size());
-    for (const auto& [label, _] : registry.components())
-      remaining.push_back(label);
-    for (const auto& label : remaining)
-      schedule_teardown(label);
-    // Finally, bring down the filesystem.
-    scheduled_for_teardown.push_back(std::move(*filesystem));
+    });
+    std::vector<caf::actor> teardown_aux;
```

Comment:
I propose changing this algorithm ever so slightly.

1. Remove the core components from the registry in the correct order, putting them into a vector of components (e.g., named `core_components`). Demonitor them.
2. Create a vector of components by removing the remaining components from the registry (e.g., named `aux_components`). Demonitor them.
3. Remove remote actors from the `aux_components` vector.
4. Do shutdown as currently implemented: `aux_components` in parallel, `core_components` sequential after that.

The logic is the same except for the additional removal from the registry, but it probably reads a lot better.